extra/changelog.py

```diff
@@ -0,0 +1,262 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""A tool for generating changelog entries.
+See more instructions at docs/unreleased-changes/README.md
+"""
+import yaml
+import os
+import sys
+import subprocess
+import argparse
+
+BASE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+CHANGES = os.path.join(BASE, 'docs', 'unreleased-changes')
+
+# Types of changelog entries.
+TYPES = ['bugfix', 'feature', 'change', 'newplugin']
+
+# List of all built-in plugins.
+PLUGINS = ['plugin/' + os.path.splitext(x)[0]
+           for x in os.listdir(os.path.join(BASE, 'beetsplug'))
+           if not x.endswith('.pyc') and not x.startswith('_')]
+
+# Components of beets that can be affected by changes.
+CORE_COMPONENTS = ['import', 'data', 'docs', 'dev']
+COMPONENTS = CORE_COMPONENTS + PLUGINS
+
+
+class ChangelogEntry(object):
+    def __init__(self, title, pull_request, type, username=None,
+                 components=None, related=None):
+        self.title = title
+        self.pr = pull_request
+        self.type = type
+        self.username = username
+        self.components = components or []
+        self.related = related or []
+
+    def as_dict(self):
+        entry = {
+            'title': self.title,
+            'pull_request': self.pr,
+            'type': self.type,
+        }
+        if username:
+            entry['username'] = username
+        if components:
+            entry['components'] = components
+        return entry
+
+    def write(self, path):
+        with open(path, 'wb') as entry_file:
+            yaml.safe_dump(self.as_dict(), entry_file, encoding='utf-8',
+                           explicit_start=True)
+
+    def print(self):
+        print(yaml.safe_dump(self.as_dict(), explicit_start=True))
+
+    def affects(self):
+        core = [x for x in sorted(self.components) if '/' not in x]
+        plugins = [':doc:`/plugins/{}`'.format(x.split('/')[1])
+                   for x in sorted(self.components)
+                   if x.startswith('plugin/')]
+        return core + plugins
+
+    def issues(self):
+        return [':bug:`{}`'.format(i) for i in [self.pr] + self.related]
+
+
+def guess_name():
+    log_entry = subprocess.Popen(['git', 'rev-parse', '--abbrev-ref', 'HEAD'],
+                                 stdout=subprocess.PIPE).stdout.read()
+    return log_entry.rstrip().decode('utf-8')
+
+
+def guess_title():
+    log_entry = subprocess.Popen(['git', 'log', '--format=%s', '-1'],
+                                 stdout=subprocess.PIPE).stdout.read()
+    return log_entry.rstrip().decode('utf-8')
+
+
+def guess_username():
+    return os.getenv(
+            'GITHUB_USER',
+            subprocess.Popen(
+                ['git', 'config', '--global', 'github.user'],
```

Comment:
The Gitlab changelog tool considers `git config user.name`, I think it would be nice to have this as an additional fallback (personally, I usually have that variable set, but not `github.user`). Is there a reason to pass `--global`? Without it, global settings will still be read, but can be overridden per-repository.

Reply:
I think `user.name` is often set to the developer's name as opposed to their GitHub username (e.g. mine is set to "Jack Wilsdon"), which isn't what is wanted here (the username ends up being appended to `https://github.com/` to create a link to their profile).

Reply:
Yes this was exactly the reason. I only wanted to collect the username rather than real name based on what was done in the past. I guess we could fall back on the real name and use a heuristic like "if it contains a space then don't expand it to a URL". It seemed simpler just to ask the user to type the username and let them know how to pre-fill it in future.

Yes it seems that the more official GitHub tools get the username from the auth token now so they don't tend to need a config item. I think it's more likely that people would set the env variable than use the git config anyway in our case. Agreed on the point about `--global`.

Reply:
Alright, sorry about the fuss. To be honest, I hadn't really checked how that value is used.